src/lib/frontend/frontend.ml

```diff
@@ -288,109 +328,157 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
+             hdist = (if is_hyp then 0 else -1);
+             trigger_depth = max_int;
+             nb_reductions = 0;
+             age=0;
+             lem=None;
+             mf=mf;
+             gf=false;
+             from_terms = [];
+             theory_elim = true;
+            }
+            expl
+        | `Unsat -> env.sat_env
+      in
+      {env with sat_env; expl}
 
-      | Query(n, f, sort) ->
-        let dep =
-          match consistent with
-          | `Sat _ | `Unknown _ ->
-            let dep' = SAT.unsat env
-                {E.ff=f;
-                 origin_name = n;
-                 hdist = -1;
-                 gdist = 0;
-                 trigger_depth = max_int;
-                 nb_reductions = 0;
-                 age=0;
-                 lem=None;
-                 mf=(sort != Check);
-                 gf=true;
-                 from_terms = [];
-                 theory_elim = true;
-                } in
-            Ex.union dep' dep
-          | `Unsat -> dep
+  let internal_pred_def ?(loc = Loc.dummy) env (name, f) =
+    if unused_context name env.used_context then
+      env
+    else
+      let expl = mk_root_dep name f loc in
+      let sat_env = SAT.pred_def env.sat_env f name expl loc in
+      {env with sat_env; expl}
+
+  let internal_query ?(loc = Loc.dummy) env (n, f, sort) =
+    ignore loc;
+    let expl =
+      match env.res with
+      | `Sat _ | `Unknown _ ->
+        let expl' = SAT.unsat env.sat_env
+            {E.ff=f;
+             origin_name = n;
+             hdist = -1;
+             gdist = 0;
+             trigger_depth = max_int;
+             nb_reductions = 0;
+             age=0;
+             lem=None;
+             mf=(sort != Ty.Check);
+             gf=true;
+             from_terms = [];
+             theory_elim = true;
+            }
         in
-        if get_debug_unsat_core () then check_produced_unsat_core dep;
-        if get_save_used_context () then output_used_context n dep;
-        print_status (Unsat (d, dep)) (Steps.get_steps ());
-        env, `Unsat, dep
+        Ex.union expl' env.expl
+      | `Unsat -> env.expl
+    in
+    if get_debug_unsat_core () then check_produced_unsat_core expl;
+    if get_save_used_context () then output_used_context n expl;
+    {env with res = `Unsat; expl}
+
+  let internal_th_assume
+      ?(loc = Loc.dummy)
+      env
+      ({ Expr.ax_name; Expr.ax_form ; _ } as th_elt) =
+    if unused_context ax_name env.used_context then
+      env
+    else
+      match env.res with
+      | `Sat _ | `Unknown _ ->
+        let expl = mk_root_dep ax_name ax_form loc in
+        let sat_env = SAT.assume_th_elt env.sat_env th_elt expl in
+        {env with sat_env; expl}
+      | `Unsat -> env
+
+  let handle_sat_exn f ?loc env x =
+    try f ?loc env x with
+    | SAT.Sat t -> {env with res = `Sat t}
+    | SAT.Unsat expl -> {env with res = `Unsat; expl = Ex.union expl env.expl}
+    | SAT.I_dont_know t -> {env with res = `Unknown t}
```

Comment:
No case for `Util.Timeout`? We raise it again in `process_decl` anyways but it could deserve at least a mention in a comment that this function is still expected to raise `Util.Timeout`.